src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+        public async Task<Stream> DialAsync(Peer peer,
+            string protocol,
+            CancellationToken cancel = default)
+        {
+            peer = RegisterPeer(peer);
+
+            // Get a connection and then a muxer to the peer.
+            var connection = await ConnectAsync(peer, cancel).ConfigureAwait(false);
+            var muxer = await connection.MuxerEstablished.Task.ConfigureAwait(false);
+
+            // Create a new stream for the peer protocol.
+            var stream = await muxer.CreateStreamAsync(protocol, cancel).ConfigureAwait(false);
+            try
+            {
+                await connection.EstablishProtocolAsync("/multistream/", stream, cancel).ConfigureAwait(false);
+
+                await Message.WriteAsync(protocol, stream, cancel).ConfigureAwait(false);
+                var result = await Message.ReadStringAsync(stream, cancel).ConfigureAwait(false);
+                if (result != protocol) throw new Exception($"Protocol '{protocol}' not supported by '{peer}'.");
+
+                return stream;
+            }
+            catch (Exception)
+            {
+                stream?.DisposeAsync();
+                throw;
+            }
+        }
+
+        /// <summary>
+        ///   Establish a duplex stream between the local and remote peer.
+        /// </summary>
+        /// <param name="remote"></param>
+        /// <param name="addrs"></param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        private async Task<PeerConnection> DialAsync(Peer remote,
+            IEnumerable<MultiAddress> addrs,
+            CancellationToken cancel)
+        {
+            _log.Debug($"Dialing {remote}");
+
+            if (remote == LocalPeer) throw new Exception("Cannot dial self.");
+
+            // If no addresses, then ask peer routing.
+            var multiAddresses = addrs.ToList();
+            if (Router != null && !multiAddresses.Any())
+            {
+                var found = await Router.FindPeerAsync(remote.Id, cancel).ConfigureAwait(false);
+                addrs = found.Addresses;
+                remote.Addresses = addrs;
+            }
+
+            // Get the addresses we can use to dial the remote.  Filter
+            // out any addresses (ip and port) we are listening on.
+            var blackList = _listeners.Keys
+               .Select(a => a.WithoutPeerId())
+               .ToArray();
+            
+            var possibleAddresses =
+                (await Task.WhenAll(multiAddresses.Select(a => a.ResolveAsync(cancel))).ConfigureAwait(false))
+               .SelectMany(a => a)
+               .Where(a => !blackList.Contains(a.WithoutPeerId()))
+               .Select(a => a.WithPeerId(remote.Id))
+               .Distinct()
+               .ToArray();
+               
+            if (possibleAddresses.Length == 0)
+            {
+                throw new Exception($"{remote} has no known or reachable address.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)